Add ldap storage

paulfariello opened the pull request:
Current implementation is read only.
Not sure if ldap storage is supposed to do vcard conversion.

I haven't added tests nor documentation but will if @pimutils thinks this patch is meaningful.

Other enhancement would be to allow for attributes configuration.

---

edit by @untitaker: Fix #283

untitaker commented:
Cool! This looks reasonable, but:

* for VCard conversion please use https://github.com/eventable/vobject instead of your own thing, which actually has some bugs with values containing backslashes or newlines. This will also make it possible to parse VCards for uploading.
* Please specifiy `ldap` and `vobject` as dependencies here: https://github.com/pimutils/vdirsyncer/blob/master/setup.py#L69 For example, if I run `pip install vdirsyncer[remotestorage]`, `requests-oauthlib` will be installed. `ldap` and `vobject` should be installed when I run `pip install vdirsyncer[ldap]`.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,84 @@
+
+class LDAPStorage(Storage):
+
+    __doc__ = '''
```
When implementing upload and update, we have to make it very clear which properties are even parsed and which ones are discarded.

untitaker replied:
Also the `__doc__` is only necessary if we need to include variables from other strings.

paulfariello replied:
Thanks for the quick reply.

I'm now using vobject which is clearly cleaner. Thanks for the hint.
I've also added support for title and fax.

I changed the default filter to one that make sense since `vobject`
clearly states that `F` and `FN` are mandatory fields. On some dirty
ldap directory `sn` or `givenName` could be missing and I'm not sure how to
handle it cleanly without filtering it. Maybe just provide the available
`family` or `givenname` for `N`.

About tests, I'm not sure how we could test it against a real ldap
server. I don't know any open ldap server but I can check.

Note that the ldap attributes I'm using are from Active Directory. Any
other ldap directory could also use it but this is not guaranteed. The
most problematic attribute is `whenChanged`. I've never seen it on other
classic ldap directory and it's used for etag.

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented:
And yes, please add tests for this, this is very meaningful! Do you have an idea how to do this? I.e., is there a simple ldap server we could use?

paulfariello commented:
Done.

On Tue, Sep 13, 2016 at 07:55:19AM -0700, Markus Unterwaditzer wrote:
> Cool! This looks reasonable, but:
> 
> * for VCard conversion please use https://github.com/eventable/vobject instead of your own thing, which actually has some bugs with values containing backslashes or newlines. This will also make it possible to parse VCards for uploading.
> * Please specifiy `ldap` and `vobject` as dependencies here: https://github.com/pimutils/vdirsyncer/blob/master/setup.py#L69 For example, if I run `pip install vdirsyncer[remotestorage]`, `requests-oauthlib` will be installed. `ldap` and `vobject` should be installed when I run `pip install vdirsyncer[ldap]`.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-246709122

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

paulfariello commented:
Thanks for the quick reply.

I'm now using vobject which is clearly cleaner. Thanks for the hint.
I've also added support for title and fax.

I changed the default filter to one that make sense since `vobject`
clearly states that `F` and `FN` are mandatory fields. On some dirty
ldap directory `sn` or `givenName` could be missing and I'm not sure how
to
handle it cleanly without filtering it. Maybe just provide the available
`family` or `givenname` for `N`.

About tests, I'm not sure how we could test it against a real ldap
server. I don't know any open ldap server but I can check.

Note that the ldap attributes I'm using are from Active Directory. Any
other ldap directory could also use it but this is not guaranteed. The
most problematic attribute is `whenChanged`. I've never seen it on other
classic ldap directory and it's used for etag.

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented:
It appears that this would be appropriate for testing: https://pypi.python.org/pypi/mockldap

See `tests/storage/test_filesystem.py` for how to write tests for the LDAP backend. In particular if you subclass `StorageTests` your storage will already be thoroughly tested, but `get_storage_args` has to be implemented.

untitaker commented:
> Note that the ldap attributes I'm using are from Active Directory. Any other ldap directory could also use it but this is not guaranteed. The most problematic attribute is `whenChanged`. I've never seen it on other classic ldap directory and it's used for etag.

Yeah I think we should make as much as possible configurable there, and also provide examples for other major LDAP providers.

untitaker commented:
Regarding `whenChanged`, if it's not there, you'll probably have to download the item and use `item.hash` or something similar.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,97 @@
+        super(LDAPStorage, self).__init__(**kwargs)
+        self.search_base = search_base
+        self.filter = filter
+        self.server = ldap3.Server(uri)
```
I think you could just use `ldap.initialize` here, no?

untitaker replied:
Ah, I see, you're using https://github.com/cannatag/ldap3, but I was reading the docs for https://github.com/pyldap/pyldap

Could you switch to the latter? This would make testing easier as `mockldap` appears to support `pyldap` only.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,97 @@
+    fileext = '.vcf'
+    item_mimetype = 'text/vcard'
+
+    def __init__(self, uri=None, search_base=None, bind=None, password=None,
```
Why are all of these None by default? The ones that can't be optional shouldn't be.

paulfariello replied:
I added a default value for uri to 'ldap://localhost'.
And try to get default value for base dn from server info.

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,97 @@
+        self.filter = filter
+        self.server = ldap3.Server(uri)
+        if bind:
+            self.conn = ldap3.Connection(self.server, user=bind,
```
What is the appropriate term, `user` or `bind`?

paulfariello replied:
"Bind dn" is the appropriate term. See ldapsearch(1) for example.
You could bind to a ldap directory with a dn that is not a user.

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented:
Am I correct to assume that this storage type does not support multiple addressbooks?

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,97 @@
+
+class LDAPStorage(Storage):
+    '''
+    :param uri: LDAP URI
```
Why URI instead of URL?

paulfariello replied:
URI is the right term for this since it can be an URL, a simple ip or a
hostname.

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker replied:
An IP or hostname is not a URI. URIs always have a scheme. https://tools.ietf.org/html/rfc3986#section-1.1.1

untitaker commented:
I added a LDAP testcase that should work if you use `pyldap` instead of `ldap3`.

paulfariello commented:
Thanks for the review. I'll fix today.

But first I have two questions:

- how am I supposed to run the test suites ? I'm running make -e test but got a bunch of errors
- how did you managed to push a commit on my fork ?

untitaker commented:
You need to run "make install-test" as well. In the documentation there is a page "Contributing" that goes into details.

Regarding how my commit landed on your fork, see https://github.com/blog/2247-improving-collaboration-with-forks

On 14 September 2016 09:23:24 CEST, Paul Fariello <notifications@github.com> wrote:
>Thanks for the review. I'll fix today.
>
>But first I have two questions:
>
>- how am I supposed to run the test suites ? I'm running make -e test
>but got a bunch of errors
>- how did you managed to push a commit on my fork ?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-246928119

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

paulfariello commented:
I've been trying to switch to pyldap but it doesn't natively support generator for ldap result. It makes it more complicated to cleanly integrate it.
Meanwhile I just discovered that ldap3 has a mocking component. I'll try to use it instead.

paulfariello commented:
Yes

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

paulfariello commented:
Tests with ldap3 mocking are still WIP.

untitaker commented:
Ah, you need to add the new dependencies to `test-requirements.txt` as well.

paulfariello commented:
Ok you are right, I'll update the code
-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

paulfariello commented:
Should I add ldap3 in test-requirements.txt even if it's in setup.py ? 
-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented:
Yes because it's a optional requirement and otherwise not installed. This
situation needs a small refactor but not for now...

On Wed, Sep 14, 2016 at 01:57:56PM -0700, Paul Fariello wrote:
> Should I add ldap3 in test-requirements.txt even if it's in setup.py ? 
> -- 
> Paul Fariello
> 
> PGP: 0x672CDD2031AAF49B
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-247151003

paulfariello commented:
Is it possible to use `StorageTests` with read only storage ?

untitaker commented:
No, unfortunately not. All it does is uploading items and then checking whether those can be downloaded again.


On 22 September 2016 09:00:46 CEST, Paul Fariello <notifications@github.com> wrote:
>Is it possible to use `StorageTests` with read only storage ?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-248827595

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

untitaker commented:
Are you interested in implementing `upload` and `update`? That would really make testing a lot easier (and I do think that is necessary).

paulfariello commented:
Main problem is that I'm not sure where to insert user on ldap directory and what fields are required.

But I'll try to add some basic update and upload. Anyway most ldap directory are read only for normal users.

Le 23 septembre 2016 10:58:51 GMT+02:00, Markus Unterwaditzer <notifications@github.com> a écrit :
>Are you interested in implementing `upload` and `update`? That would
>really make testing a lot easier (and I do think that is necessary).
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-249139442

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

untitaker commented:
I see. In that case it might make sense to write upload and update methods that just upload the data where the other storage methods can find it again, just for testing, but actually force the user to use read-only mode (I'd write this part)

On 23 September 2016 18:13:45 CEST, Paul Fariello <notifications@github.com> wrote:
>Main problem is that I'm not sure where to insert user on ldap
>directory and what fields are required.
>
>But I'll try to add some basic update and upload. Anyway most ldap
>directory are read only for normal users.
>
>Le 23 septembre 2016 10:58:51 GMT+02:00, Markus Unterwaditzer
><notifications@github.com> a écrit :
>>Are you interested in implementing `upload` and `update`? That would
>>really make testing a lot easier (and I do think that is necessary).
>>
>>-- 
>>You are receiving this because you authored the thread.
>>Reply to this email directly or view it on GitHub:
>>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-249139442
>
>-- 
>Paul Fariello
>
>PGP: 0x672CDD2031AAF49B
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-249235546

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

untitaker commented:
@paulfariello BTW let me know if I can help with anything or if anything is unclear.

untitaker commented:
@paulfariello Any update on this? Adding tests is really important! Are you using this already for yourself?

paulfariello commented:
Sorry I've been quite busy lately. I'll try to add tests by the end of next week.

Le 8 octobre 2016 14:24:04 GMT+02:00, Markus Unterwaditzer <notifications@github.com> a écrit :
>@paulfariello Any update on this? Adding tests is really important! Are
>you using this already for yourself?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/vdirsyncer/pull/500#issuecomment-252421803

-- 
Paul Fariello

PGP: 0x672CDD2031AAF49B

paulfariello commented:
I tried to add upload and update method. But I'm facing an issue with VCALENDAR. What should I do with it ? They doesn't really belong to an LDAP directory.

untitaker commented:
Fantastic, please push what you have, I'll do the changes such that VCALENDARs are not used in the tests.

paulfariello commented:
My last push is just a first attempt on update and upload implementation. It push to mocked server.
I also have an issue with start tls on mocked server.

untitaker commented:
Are `ldap://`-urls SSL by default?

untitaker commented:
I believe I've fixed the issue with ssl, let's just expect that the connection is fully set up if it's passed into the storage. I think ldap doesn't deal with SSL very well when mocked, but in our case SSL is not of any use anyway.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,118 @@
+
+    def upload(self, item):
+        vcard = vobject.readOne(item.raw)
+        self.conn.strategy.add_entry('cn={},ou=test,o=lab'.format(vcard.fn),
```
Upload needs to return the assigned href and etag. It appears that your href is `dn`, but the etag has to be fetched like in `list` and `get`.

Also `add_entry` returns a boolean that indicates whether the entry has been added, you should check for that.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,118 @@
+            raise exceptions.NotFoundError(href)
+
+        entry = self.conn.entries[0]
+        etag = str(entry.whenChanged)
```
Here you have no fallback from `whenChanged` to `item.hash`. I suggest a separate helper function (outside of the storage class) for this.

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,118 @@
+        vo = vcard.add('n')
+        vo.value = vobject.vcard.Name(family=str(entry.sn),
+                                      given=str(entry.givenName))
+        if getattr(entry, 'telephoneNumber', None):
```
This conversion from entry to item could be its own function (outside of storage class)

untitaker commented on vdirsyncer/storage/ldap.py:
```diff
@@ -0,0 +1,118 @@
+
+    def update(self, href, item, etag):
+        vcard = vobject.readOne(item.raw)
+        self.conn.strategy.add_entry(href, vcard)
```
`add_entry` will do nothing if the item already exists, but if I call `update`, the item always exists. `update` is also expected to fail if the passed `etag` does not match the `etag` of the item on the server.

I think `modify` or how it's called is more appropriate than `add_entry`, but I'm not sure since I couldn't find much docs :(

untitaker commented:
Also please [contact](https://vdirsyncer.pimutils.org/en/stable/contact.html) me if there are any questions at all, PM, direct email, whatever too.

untitaker commented:
@paulfariello Ping!

untitaker commented:
Closing this due to inactivity, but for reference, this is a good start for a successor PR.
